File: lld/week-02-03-patterns/builder/go/builder_test.go
package builder

import (
	"strings"
	"testing"
)

func TestFluentChainBuildsExpectedRequest(t *testing.T) {
	req, err := NewHttpRequestBuilder().
		Method("POST").
		URL("https://api.example.com/orders").
		Header("Content-Type", "application/json").
		Header("Authorization", "Bearer token123").
		Body(`{"item":"widget"}`).
		Build()
	if err != nil {
		t.Fatalf("unexpected err: %v", err)
	}

	if req.Method() != "POST" {
		t.Fatalf("expected POST, got %s", req.Method())
	}
	if req.URL() != "https://api.example.com/orders" {
		t.Fatalf("unexpected URL: %s", req.URL())
	}
	if v, ok := req.Header("Content-Type"); !ok || v != "application/json" {
		t.Fatalf("unexpected Content-Type header: %s, %v", v, ok)
	}
	if req.Body() != `{"item":"widget"}` {
		t.Fatalf("unexpected body: %s", req.Body())
	}
}

func TestDefaultMethodIsGet(t *testing.T) {
	req, err := NewHttpRequestBuilder().URL("https://example.com").Build()
	if err != nil {
		t.Fatalf("unexpected err: %v", err)
	}
	if req.Method() != "GET" {
		t.Fatalf("expected default GET, got %s", req.Method())
	}
}

func TestBuildFailsWithoutURL(t *testing.T) {
	_, err := NewHttpRequestBuilder().Method("GET").Build()
	if err != ErrMissingURL {
		t.Fatalf("expected ErrMissingURL, got %v", err)
	}
}

func TestBuiltRequestIsImmutableFromBuilderReuse(t *testing.T) {
	b := NewHttpRequestBuilder().URL("https://example.com").Header("X-A", "1")
	first, err := b.Build()
	if err != nil {
		t.Fatalf("unexpected err: %v", err)
	}

	b.Header("X-A", "2").Header("X-B", "new")
	second, err := b.Build()
	if err != nil {
		t.Fatalf("unexpected err: %v", err)
	}

	if v, _ := first.Header("X-A"); v != "1" {
		t.Fatalf("expected first request's X-A to stay 1, got %s", v)
	}
	if _, ok := first.Header("X-B"); ok {
		t.Fatalf("first request should not see header added after it was built")
	}
	if v, _ := second.Header("X-A"); v != "2" {
		t.Fatalf("expected second request's X-A to be 2, got %s", v)
	}
}

func TestStringRendersDeterministicHeaderOrder(t *testing.T) {
	req, _ := NewHttpRequestBuilder().
		Method("GET").
		URL("https://example.com").
		Header("Z-Header", "z").
		Header("A-Header", "a").
		Build()

	s := req.String()
	if strings.Index(s, "A-Header") > strings.Index(s, "Z-Header") {
		t.Fatalf("expected headers sorted alphabetically, got:\n%s", s)
	}
}

File: lld/week-02-03-patterns/builder/java/src/BuilderTest.java
/**
 * Plain assert-based test runner (no JUnit dependency needed).
 * Run via `java -cp out Main` (invoked from Main) or `java -cp out BuilderTest` directly.
 */
public class BuilderTest {

    public static void main(String[] args) {
        runAll();
    }

    public static void runAll() {
        testFluentChainBuildsExpectedRequest();
        testDefaultMethodIsGet();
        testBuildFailsWithoutUrl();
        testBuiltRequestIsImmutableFromBuilderReuse();
        System.out.println("All BuilderTest cases passed.");
    }

    private static void testFluentChainBuildsExpectedRequest() {
        HttpRequest req = new HttpRequestBuilder()
                .method("POST")
                .url("https://api.example.com/orders")
                .header("Content-Type", "application/json")
                .body("{\"item\":\"widget\"}")
                .build();

        assertEquals("POST", req.getMethod(), "method");
        assertEquals("https://api.example.com/orders", req.getUrl(), "url");
        assertEquals("application/json", req.getHeader("Content-Type"), "content-type header");
        assertEquals("{\"item\":\"widget\"}", req.getBody(), "body");
    }

    private static void testDefaultMethodIsGet() {
        HttpRequest req = new HttpRequestBuilder().url("https://example.com").build();
        assertEquals("GET", req.getMethod(), "default method");
    }

    private static void testBuildFailsWithoutUrl() {
        try {
            new HttpRequestBuilder().method("GET").build();
            throw new AssertionError("expected MissingUrlException");
        } catch (HttpRequestBuilder.MissingUrlException e) {
            // expected
        }
    }

    private static void testBuiltRequestIsImmutableFromBuilderReuse() {
        HttpRequestBuilder b = new HttpRequestBuilder().url("https://example.com").header("X-A", "1");
        HttpRequest first = b.build();

        b.header("X-A", "2").header("X-B", "new");
        HttpRequest second = b.build();

        assertEquals("1", first.getHeader("X-A"), "first request X-A unchanged");
        assertEquals(null, first.getHeader("X-B"), "first request should not see later header");
        assertEquals("2", second.getHeader("X-A"), "second request X-A updated");
    }

    private static void assertEquals(Object expected, Object actual, String label) {
        boolean equal = (expected == null) ? (actual == null) : expected.equals(actual);
        if (!equal) {
            throw new AssertionError(label + ": expected " + expected + " but got " + actual);
        }
    }
}
